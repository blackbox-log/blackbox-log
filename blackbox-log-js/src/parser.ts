import { File } from './file';
import { getWasmStr } from './str';

import type { WasmExports } from './wasm';

export type WasmInit = string | URL | Request | Response | WebAssembly.Module;

type Instance = WebAssembly.Instance & { exports: WasmExports };

export class Parser {
	static async init(init: WasmInit): Promise<Parser> {
		let instance: Instance | undefined;

		const imports = {
			main: {
				panic(len: number, ptr: number) {
					if (instance === undefined) {
						console.error('received panic before JS handler was initialized');
						return;
					}

					console.error(getWasmStr([len, ptr], instance.exports));
				},
				throw(len: number, ptr: number) {
					if (instance === undefined) {
						throw new ParseError('unknown error');
					} else {
						const message = getWasmStr([len, ptr], instance.exports);
						instance.exports.slice8_free(len, ptr);
						throw new ParseError(message);
					}
				},
			},
		};

		if (init instanceof WebAssembly.Module) {
			instance = new WebAssembly.Instance(init, imports) as Instance;
		} else {
			const response = init instanceof Response ? init : fetch(init);

			const streamed = await WebAssembly.instantiateStreaming(response, imports);
			instance = streamed.instance as Instance;
		}

		(instance.exports.set_panic_hook as () => void)();

		return new Parser(instance as { exports: WasmExports });
	}

	readonly #wasm: WasmExports;

	constructor(wasm: WebAssembly.Instance) {
		this.#wasm = wasm.exports as WasmExports;
	}

	loadFile(data: Uint8Array): File {
		return new File(this.#wasm, data);
	}
}

export class ParseError extends Error {
	constructor(message: string, options?: ErrorOptions) {
		super(message, options);

		// Maintain V8 stack trace
		// @ts-expect-error Only present on V8 and is missing from typedef
		Error.captureStackTrace?.(this, ParseError); // eslint-disable-line @typescript-eslint/no-unsafe-call

		this.name = 'ParseError';
	}
}
