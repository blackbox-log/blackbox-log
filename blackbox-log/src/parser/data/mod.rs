mod event;

use alloc::vec::Vec;

pub use self::event::Event;
use super::{FrameKind, Headers, MainFrame, ParseError, ParseResult, Reader, SlowFrame};

#[derive(Debug, Clone)]
pub struct Data {
    pub(crate) events: Vec<Event>,
    pub(crate) main_frames: Vec<(MainFrame, usize)>,
    // pub(crate) gps_frames: Vec<Frame>,
    // pub(crate) gps_home_frames: Vec<Frame>,
    pub(crate) slow_frames: Vec<SlowFrame>,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct Stats {
    pub counts: FrameCounts,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct FrameCounts {
    pub main: usize,
    pub slow: usize,
}

impl Data {
    pub(crate) fn to_stats(&self) -> Stats {
        Stats {
            counts: FrameCounts {
                main: self.main_frames.len(),
                slow: self.slow_frames.len(),
            },
        }
    }

    pub fn parse(mut data: Reader, headers: &Headers) -> ParseResult<Self> {
        let mut events = Vec::new();
        let mut main_frames = Vec::new();
        // let gps_frames = Vec::new();
        // let gps_home_frames = Vec::new();
        let mut slow_frames = Vec::new();

        slow_frames.push(headers.slow_frames.default_frame(headers));

        let mut last_kind = None;
        while let Some(kind) = data.read_u8().map(FrameKind::from_byte) {
            // TODO (rust 1.65): let-else
            let kind = if let Some(kind) = kind {
                kind
            } else {
                tracing::error!("found invalid frame byte");
                match last_kind.take() {
                    Some(FrameKind::Event) => {
                        events.pop();
                    }
                    Some(FrameKind::Intra | FrameKind::Inter) => {
                        main_frames.pop();
                    }
                    Some(FrameKind::Slow) => {
                        slow_frames.pop();
                    }
                    Some(FrameKind::Gps | FrameKind::GpsHome) | None => {}
                };

                data.skip_until_any(
                    &[
                        FrameKind::Event,
                        FrameKind::Intra,
                        FrameKind::Slow,
                        FrameKind::Gps,
                        FrameKind::GpsHome,
                    ]
                    .map(u8::from),
                );

                continue;
            };

            let result = match kind {
                FrameKind::Event => match Event::parse_into(&mut data, &mut events) {
                    Ok(event::EventKind::End) => {
                        tracing::trace!("found the end event");
                        break;
                    }
                    Ok(_) => Ok(()),
                    Err(err) => Err(err),
                },
                FrameKind::Intra | FrameKind::Inter => {
                    let frame = MainFrame::parse(&mut data, kind, &main_frames, headers);
                    frame.map(|frame| main_frames.push((frame, slow_frames.len() - 1)))
                }
                FrameKind::Gps => {
                    if let Some(ref gps) = headers.gps_frames {
                        gps.parse(&mut data, headers).map(|_| ())
                    } else {
                        tracing::error!("found GPS frame without GPS frame definition");
                        return Err(ParseError::Corrupted);
                    }
                }
                FrameKind::GpsHome => {
                    if let Some(ref gps_home) = headers.gps_home_frames {
                        gps_home.parse(&mut data, headers).map(|_| ())
                    } else {
                        tracing::error!("found GPS home frame without GPS frame definition");
                        return Err(ParseError::Corrupted);
                    }
                }
                FrameKind::Slow => headers
                    .slow_frames
                    .parse(&mut data, headers)
                    .map(|frame| slow_frames.push(frame)),
            };

            match result {
                Ok(()) => {
                    last_kind = Some(kind);
                }
                Err(ParseError::UnexpectedEof) => {
                    tracing::warn!("found unexpected end of file");
                    break;
                }
                Err(err) => return Err(err),
            }
        }

        Ok(Self {
            events,
            main_frames,
            // gps_frames,
            // gps_home_frames,
            slow_frames,
        })
    }
}
