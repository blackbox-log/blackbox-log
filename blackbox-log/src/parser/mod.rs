pub(crate) mod decode;

pub(crate) use self::decode::Encoding;

pub(crate) type InternalResult<T> = Result<T, InternalError>;

/// A recoverable error.
#[derive(Debug, Clone)]
pub(crate) enum InternalError {
    /// Found something unexpected, retry at the next byte.
    Retry,
    /// Found the end of file.
    Eof,
}

pub(crate) fn to_base_field(field: &str) -> &str {
    field.split_once('[').map_or(field, |(base, _)| base)
}
