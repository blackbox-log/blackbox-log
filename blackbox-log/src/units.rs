use alloc::vec::Vec;
use core::fmt;

use bitvec::prelude::*;

use crate::common::FirmwareKind;
use crate::parser::headers::CurrentMeterConfig;
use crate::parser::{as_signed, Headers};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Amperage {
    raw: i32,
    current_meter: CurrentMeterConfig,
}

impl Amperage {
    pub(crate) fn new(raw: u32, headers: &Headers) -> Self {
        Self {
            raw: as_signed(raw),
            current_meter: headers.current_meter.unwrap(),
        }
    }

    pub fn as_amps(&self) -> f64 {
        let milliamps = f64::from(self.raw * 3300) / 4095.;
        let milliamps = milliamps - f64::from(self.current_meter.offset);
        (milliamps * 10.) / f64::from(self.current_meter.scale)
    }
}

impl fmt::Display for Amperage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.*}", f.precision().unwrap_or(2), self.as_amps())
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Voltage {
    pub(crate) raw: u32,
    pub(crate) scale: u16,
}

impl Voltage {
    pub(crate) fn new(raw: u32, headers: &Headers) -> Self {
        Self {
            raw,
            scale: headers.vbat.unwrap().scale,
        }
    }

    pub fn as_volts(&self) -> f64 {
        f64::from(
            self.raw
                .saturating_mul(330)
                .saturating_mul(u32::from(self.scale)),
        ) / 4.095
    }
}

impl fmt::Display for Voltage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.*}", f.precision().unwrap_or(2), self.as_volts())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Acceleration {
    raw: i32,
    one_g: u16,
}

impl Acceleration {
    pub(crate) fn new(raw: u32, headers: &Headers) -> Self {
        Self {
            raw: as_signed(raw),
            one_g: headers.acceleration_1g.unwrap(),
        }
    }

    pub fn as_gs(&self) -> f64 {
        f64::from(self.raw) / f64::from(self.one_g)
    }
}

impl fmt::Display for Acceleration {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.*}", f.precision().unwrap_or(2), self.as_gs())
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Rotation(i32);

impl Rotation {
    pub(crate) const fn new(raw: u32) -> Self {
        Self(as_signed(raw))
    }

    pub fn as_degrees(&self) -> i32 {
        self.0
    }
}

impl fmt::Display for Rotation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub trait FlagSet {
    type Flag: Flag;

    fn is_set(&self, flag: Self::Flag) -> bool;
    fn as_names(&self) -> Vec<&'static str>;
}

pub trait Flag {
    fn as_name(&self) -> &'static str;
}

macro_rules! define_flag_set {
    ($set:ident, $flag_name:ident {
        $( $flag:ident : $($beta:literal)? / $($inav:literal)? ),* $(,)?
    }) => {
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub struct $set {
            firmware: FirmwareKind,
            raw: BitArray<[u32; 1], Lsb0>,
        }

        impl $set {
            pub(crate) fn new(raw: u32, firmware: FirmwareKind) -> Self {
                Self {
                    firmware,
                    raw: BitArray::new([raw]),
                }
            }
        }

        impl FlagSet for $set {
            type Flag = $flag_name;

            fn is_set(&self, flag: Self::Flag) -> bool {
                flag.to_bit(self.firmware)
                    .map_or(false, |bit| self.raw[bit])
            }

            fn as_names(&self) -> Vec<&'static str> {
                self.raw
                    .iter_ones()
                    .filter_map(|bit| Some($flag_name::from_bit(bit, self.firmware)?.as_name()))
                    .collect()
            }
        }

        impl fmt::Display for $set {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str(&self.as_names().join("|"))
            }
        }

        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub enum $flag_name {
            $( $flag ),*
        }

        impl Flag for $flag_name {
            fn as_name(&self) -> &'static str {
                match self {
                    $( Self::$flag => stringify!($flag) ),*
                }
            }
        }

        impl $flag_name {
            const fn from_bit(bit: usize, firmware: FirmwareKind) -> Option<Self> {
                match (bit, firmware) {
                    $($( ($beta, FirmwareKind::Betaflight) => Some(Self::$flag), )?)*
                    $($( ($inav, FirmwareKind::INav) => Some(Self::$flag), )?)*
                    _ => None,
                }
            }

            const fn to_bit(self, firmware: FirmwareKind) -> Option<usize> {
                match (self, firmware) {
                    $($( (Self::$flag, FirmwareKind::Betaflight) => Some($beta), )?)*
                    $($( (Self::$flag, FirmwareKind::INav) => Some($inav), )?)*
                    _ => None,
                }
            }
        }

        impl fmt::Display for $flag_name {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str(self.as_name())
            }
        }
    };
}

define_flag_set!(FlightModeSet, FlightMode {
    Angle:     1 /  0,
    Horizon:   2 /  1,
    HeadFree:  4 /  6,
    Failsafe:  6 /  9,
    Turtle:   27 / 15,

    Arm:            0 / ,
    Mag:            3 / ,
    Passthru:       5 / ,
    GpsRescue:      7 / ,
    Antigravity:    8 / ,
    HeadAdjust:     9 / ,
    CamStab:       10 / ,
    BeeperOn:      11 / ,
    LedLow:        12 / ,
    Calib:         13 / ,
    Osd:           14 / ,
    Telemetry:     15 / ,
    Servo1:        16 / ,
    Servo2:        17 / ,
    Servo3:        18 / ,
    Blackbox:      19 / ,
    Airmode:       20 / ,
    ThreeD:        21 / ,
    FpvAngleMix:   22 / ,
    BlackboxErase: 23 / ,
    Camera1:       24 / ,
    Camera2:       25 / ,
    Camera3:       26 / ,
    Prearm:        28 / ,
    BeepGpsCount:  29 / ,
    VtxPitmode:    30 / ,
    Paralyze:      31 / ,

    // User1:               32 / ,
    // User2:               33 / ,
    // User3:               34 / ,
    // User4:               35 / ,
    // PidAudio:            36 / ,
    // AcroTrainer:         37 / ,
    // VtxControlDisable:   38 / ,
    // LaunchControl:       39 / ,
    // MspOverride:         40 / ,
    // StickCommandDisable: 41 / ,
    // BeeperMute:          42 / ,

    Heading:       /  2,
    NavAltHold:    /  3,
    NavRth:        /  4,
    NavPoshold:    /  5,
    NavLaunch:     /  7,
    Manual:        /  8,
    AutoTune:      / 10,
    NavWp:         / 11,
    NavCourseHold: / 12,
    Flaperon:      / 13,
    TurnAssistant: / 14,
    Soaring:       / 16,
});

define_flag_set!(StateSet, State {
    GpsFixHome: 0 / 0,
    GpsFix:     1 / 1,
    GpsFixEver: 2 /  ,

    CalibrateMag:                 /  2,
    SmallAngle:                   /  3,
    FixedWingLegacy:              /  4,
    AntiWindup:                   /  5,
    FlaperonAvailable:            /  6,
    NavMotorStopOrIdle:           /  7,
    CompassCalibrated:            /  8,
    AccelerometerCalibrated:      /  9,
    NavCruiseBraking:             / 11,
    NavCruiseBrakingBoost:        / 12,
    NavCruiseBrakingLocked:       / 13,
    NavExtraArmingSafetyBypassed: / 14,
    AirmodeActive:                / 15,
    EscSensorEnabled:             / 16,
    Airplane:                     / 17,
    Multirotor:                   / 18,
    Rover:                        / 19,
    Boat:                         / 20,
    AltitudeControl:              / 21,
    MoveForwardOnly:              / 22,
    SetReversibleMotorsForward:   / 23,
    FwHeadingUseYaw:              / 24,
    AntiWindupDeactivated:        / 25,
    LandingDetected:              / 26,
});

define_flag_set!(FailsafePhaseSet, FailsafePhase {
    Idle:             0 / 0,
    RxLossDetected:   1 / 1,
    RxLossIdle:         / 2,
    ReturnToHome:       / 3,
    Landing:          2 / 4,
    Landed:           3 / 5,
    RxLossMonitoring: 4 / 6,
    RxLossRecovered:  5 / 7,
    GpsRescue:        6 /  ,
});
