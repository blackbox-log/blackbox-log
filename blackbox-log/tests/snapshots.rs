use std::fs;
use std::io::Read;

use blackbox_log::parser::{Event, Headers, MainValue, SlowValue, Stats};
use blackbox_log::units::FlagSet;
use blackbox_log::Log;
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

#[test]
fn snapshot() {
    insta::glob!("logs/*", |path| {
        let mut file = fs::File::open(path).unwrap();
        let mut data = Vec::new();
        file.read_to_end(&mut data).unwrap();

        let file = blackbox_log::File::new(&data);
        let logs = file
            .parse_iter()
            .map(|r| r.map(LogSnapshot::from))
            .collect::<Vec<_>>();

        insta::assert_ron_snapshot!(logs);
    });
}

#[derive(Debug, Serialize)]
struct LogSnapshot<'a> {
    headers: Headers<'a>,
    stats: Stats,
    events: Vec<Event>,
    main: FrameSnapshot,
    slow: FrameSnapshot,
}

impl<'a> From<Log<'a>> for LogSnapshot<'a> {
    fn from(log: Log<'a>) -> Self {
        let main = log
            .main_fields()
            .map(|(name, _)| name)
            .collect::<FrameSnapshot>();
        let slow = log
            .slow_fields()
            .map(|(name, _)| name)
            .collect::<FrameSnapshot>();

        let (main, slow) = log
            .iter_frames()
            .fold((main, slow), |(mut main, mut slow), frame| {
                main.update(frame.iter_main().map(main_to_int));
                slow.update(frame.iter_slow().map(slow_to_int));

                (main, slow)
            });

        Self {
            headers: log.headers().clone(),
            stats: log.stats(),
            events: log.events().to_owned(),
            main,
            slow,
        }
    }
}

#[derive(Debug, Serialize)]
struct FrameSnapshot {
    fields: Vec<FieldSnapshot>,
}

impl FrameSnapshot {
    fn update(&mut self, frame: impl Iterator<Item = i128>) {
        for (field, value) in self.fields.iter_mut().zip(frame) {
            field.update(value);
        }
    }
}

impl<T> FromIterator<T> for FrameSnapshot
where
    T: Into<String>,
{
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
        Self {
            fields: iter.into_iter().map(FieldSnapshot::new).collect(),
        }
    }
}

#[derive(Debug)]
struct FieldSnapshot {
    name: String,
    min: i128,
    max: i128,
    seen: Vec<i128>,
    histogram: [i128; 16],
}

impl FieldSnapshot {
    fn new<T: Into<String>>(name: T) -> Self {
        Self {
            name: name.into(),
            min: 0,
            max: 0,
            seen: Vec::new(),
            histogram: [0; 16],
        }
    }

    fn update(&mut self, value: i128) {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }

        // Insert new value if not in seen, keeping seen sorted
        let index = self.seen.partition_point(|&x| x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }

        // Group into buckets using the bottom bits since those vary the most
        let index = (value.unsigned_abs() % 16) as usize;
        self.histogram[index] += 1;
    }
}

impl Serialize for FieldSnapshot {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let unique = self.seen.len();

        let mut state =
            serializer.serialize_struct("FieldSnapshot", if unique > 1 { 5 } else { 4 })?;

        state.serialize_field("name", &self.name)?;
        state.serialize_field("min", &self.min)?;
        state.serialize_field("max", &self.max)?;
        state.serialize_field("unique", &unique)?;

        if unique > 1 {
            state.serialize_field("histogram", &self.histogram)?;
        }

        state.end()
    }
}

fn slow_to_int(value: SlowValue) -> i128 {
    match value {
        SlowValue::FlightMode(x) => x.as_raw().into(),
        SlowValue::Unsigned(x) => x.into(),
        SlowValue::Signed(x) => x.into(),
    }
}

fn main_to_int(value: MainValue) -> i128 {
    match value {
        MainValue::FrameTime(x) => x.into(),
        MainValue::Amperage(x) => x.as_raw().into(),
        MainValue::Voltage(x) => x.as_raw().into(),
        MainValue::Acceleration(x) => x.as_raw().into(),
        MainValue::Rotation(x) => x.as_raw().into(),
        MainValue::Unsigned(x) => x.into(),
        MainValue::Signed(x) => x.into(),
    }
}
