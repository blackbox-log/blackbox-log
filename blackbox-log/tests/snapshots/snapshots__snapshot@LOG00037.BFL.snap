---
source: blackbox-log/tests/snapshots.rs
expression: logs
input_file: blackbox-log/tests/logs/LOG00037.BFL
---
[
  Ok(LogSnapshot(
    headers: Headers(
      version: V2,
      firmware_revision: "Betaflight 4.2.0 (8f2d21460) STM32F745",
      firmware_kind: Betaflight,
      board_info: "HBRO KAKUTEF7",
      craft_name: "AR8",
      vbat_reference: 2277,
      vbat_scale: 110,
      current_meter: CurrentMeterConfig(
        offset: 0,
        scale: 275,
      ),
      acceleration_1g: 2048,
      gyro_scale: 1.0,
      min_throttle: 1070,
      motor_output_range: MotorOutputRange(158, 2047),
    ),
    stats: Stats(
      counts: FrameCounts(
        main: 16774,
        slow: 4,
      ),
    ),
    events: [
      SyncBeep(451840837),
      Disarm(4),
      End,
    ],
    main: FrameSnapshot(
      fields: [
        FieldSnapshot(
          name: "loopIteration",
          min: 0,
          max: 134149,
          unique: 16774,
          histogram: (1049, 1049, 1049, 1049, 1049, 1049, 1048, 1048, 1048, 1048, 1048, 1048, 1048, 1048, 1048, 1048),
        ),
        FieldSnapshot(
          name: "time",
          min: 0,
          max: 469230773,
          unique: 16774,
          histogram: (1070, 1082, 1113, 945, 1091, 677, 1378, 1050, 1048, 1050, 1119, 951, 1102, 684, 1362, 1052),
        ),
        FieldSnapshot(
          name: "axisP[0]",
          min: -49,
          max: 80,
          unique: 89,
          histogram: (3977, 3227, 2530, 1824, 1341, 960, 672, 541, 367, 292, 243, 241, 192, 184, 117, 66),
        ),
        FieldSnapshot(
          name: "axisP[1]",
          min: -583,
          max: 284,
          unique: 198,
          histogram: (3662, 2960, 2170, 1506, 1171, 918, 826, 697, 614, 532, 421, 398, 316, 223, 181, 179),
        ),
        FieldSnapshot(
          name: "axisP[2]",
          min: -73,
          max: 23,
          unique: 72,
          histogram: (6030, 4391, 2658, 1541, 916, 496, 243, 160, 95, 69, 49, 42, 32, 14, 22, 16),
        ),
        FieldSnapshot(
          name: "axisI[0]",
          min: -50,
          max: 11,
          unique: 62,
          histogram: (1888, 864, 1203, 1858, 1494, 1409, 1772, 1633, 1183, 799, 506, 377, 347, 416, 483, 542),
        ),
        FieldSnapshot(
          name: "axisI[1]",
          min: -251,
          max: 92,
          unique: 195,
          histogram: (1567, 839, 1346, 1352, 1610, 2571, 1943, 1133, 1053, 942, 765, 460, 403, 306, 258, 226),
        ),
        FieldSnapshot(
          name: "axisI[2]",
          min: -61,
          max: 44,
          unique: 101,
          histogram: (2713, 1770, 2205, 1936, 1969, 1962, 1453, 906, 614, 476, 231, 146, 131, 104, 64, 94),
        ),
        FieldSnapshot(
          name: "axisD[0]",
          min: -122,
          max: 192,
          unique: 106,
          histogram: (3422, 3001, 2709, 2042, 1647, 1219, 864, 605, 405, 274, 181, 124, 95, 72, 60, 54),
        ),
        FieldSnapshot(
          name: "axisD[1]",
          min: -893,
          max: 526,
          unique: 140,
          histogram: (3571, 3196, 2698, 2045, 1537, 1019, 742, 592, 384, 265, 225, 163, 137, 85, 69, 46),
        ),
        FieldSnapshot(
          name: "axisF[0]",
          min: -1,
          max: 3,
          unique: 5,
          histogram: (16419, 312, 23, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
        FieldSnapshot(
          name: "axisF[1]",
          min: -2,
          max: 2,
          unique: 5,
          histogram: (16269, 436, 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
        FieldSnapshot(
          name: "axisF[2]",
          min: -7,
          max: 1,
          unique: 9,
          histogram: (16539, 190, 9, 5, 4, 7, 6, 14, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
        FieldSnapshot(
          name: "rcCommand[0]",
          min: -85,
          max: 98,
          unique: 184,
          histogram: (7414, 1242, 544, 800, 613, 687, 543, 740, 459, 576, 617, 375, 495, 652, 456, 561),
        ),
        FieldSnapshot(
          name: "rcCommand[1]",
          min: -76,
          max: 111,
          unique: 188,
          histogram: (985, 1079, 802, 6789, 1319, 736, 625, 703, 404, 426, 512, 361, 500, 675, 405, 453),
        ),
        FieldSnapshot(
          name: "rcCommand[2]",
          min: -96,
          max: 78,
          unique: 175,
          histogram: (1443, 4241, 1232, 1963, 1109, 883, 532, 1212, 382, 529, 531, 402, 627, 880, 407, 401),
        ),
        FieldSnapshot(
          name: "rcCommand[3]",
          min: 0,
          max: 1948,
          unique: 754,
          histogram: (950, 850, 794, 582, 880, 1324, 1449, 765, 2428, 798, 642, 1286, 1288, 986, 1004, 748),
        ),
        FieldSnapshot(
          name: "setpoint[0]",
          min: -27,
          max: 31,
          unique: 59,
          histogram: (7432, 1193, 766, 794, 914, 1184, 696, 694, 702, 494, 374, 452, 345, 197, 292, 245),
        ),
        FieldSnapshot(
          name: "setpoint[1]",
          min: -23,
          max: 36,
          unique: 60,
          histogram: (1359, 8165, 935, 851, 907, 857, 890, 535, 486, 322, 264, 225, 225, 218, 313, 222),
        ),
        FieldSnapshot(
          name: "setpoint[2]",
          min: -35,
          max: 28,
          unique: 64,
          histogram: (4666, 3142, 1700, 827, 1018, 856, 562, 972, 673, 351, 248, 497, 393, 253, 283, 333),
        ),
        FieldSnapshot(
          name: "setpoint[3]",
          min: 0,
          max: 965,
          unique: 729,
          histogram: (2417, 759, 774, 1271, 1280, 956, 958, 686, 926, 823, 796, 690, 911, 1293, 1345, 889),
        ),
        FieldSnapshot(
          name: "vbatLatest",
          min: 0,
          max: 2281,
          unique: 140,
          histogram: (849, 989, 983, 1145, 1577, 1266, 1006, 1044, 927, 987, 1067, 989, 1066, 1002, 770, 1107),
        ),
        FieldSnapshot(
          name: "amperageLatest",
          min: 0,
          max: 11978,
          unique: 591,
          histogram: (1838, 666, 1678, 787, 556, 1603, 669, 569, 1617, 692, 1916, 968, 689, 1265, 530, 731),
        ),
        FieldSnapshot(
          name: "magADC[0]",
          min: -740,
          max: 292,
          unique: 117,
          histogram: (1282, 791, 1282, 1677, 984, 786, 1185, 493, 1084, 591, 592, 1480, 1427, 984, 852, 1284),
        ),
        FieldSnapshot(
          name: "magADC[1]",
          min: -66,
          max: 947,
          unique: 126,
          histogram: (1480, 395, 985, 888, 1382, 1775, 393, 982, 1679, 951, 1476, 987, 689, 1184, 639, 889),
        ),
        FieldSnapshot(
          name: "magADC[2]",
          min: 0,
          max: 2576,
          unique: 123,
          histogram: (1187, 1281, 1084, 788, 493, 690, 1084, 1133, 885, 1085, 1149, 1579, 1084, 1182, 788, 1282),
        ),
        FieldSnapshot(
          name: "BaroAlt",
          min: -656,
          max: 289,
          unique: 255,
          histogram: (914, 419, 1376, 979, 1014, 716, 1506, 2241, 690, 1057, 689, 839, 1078, 568, 1652, 1036),
        ),
        FieldSnapshot(
          name: "rssi",
          min: 0,
          max: 1023,
          unique: 1,
        ),
        FieldSnapshot(
          name: "gyroADC[0]",
          min: -30,
          max: 37,
          unique: 68,
          histogram: (3820, 3984, 1450, 956, 904, 827, 664, 694, 672, 574, 486, 512, 415, 386, 236, 194),
        ),
        FieldSnapshot(
          name: "gyroADC[1]",
          min: -120,
          max: 197,
          unique: 136,
          histogram: (3222, 4965, 1926, 1212, 953, 853, 674, 705, 501, 361, 324, 194, 254, 317, 134, 179),
        ),
        FieldSnapshot(
          name: "gyroADC[2]",
          min: -43,
          max: 28,
          unique: 72,
          histogram: (3436, 3829, 1941, 1067, 995, 946, 739, 901, 532, 342, 394, 391, 383, 347, 280, 251),
        ),
        FieldSnapshot(
          name: "accSmooth[0]",
          min: -1496,
          max: 1183,
          unique: 604,
          histogram: (1039, 1172, 1200, 1178, 1133, 1188, 1015, 1077, 1050, 963, 1003, 1008, 981, 939, 926, 902),
        ),
        FieldSnapshot(
          name: "accSmooth[1]",
          min: -864,
          max: 71,
          unique: 380,
          histogram: (986, 983, 1037, 1006, 1014, 1064, 1021, 1051, 1051, 1070, 1049, 1096, 1100, 1106, 1160, 980),
        ),
        FieldSnapshot(
          name: "accSmooth[2]",
          min: 0,
          max: 7547,
          unique: 2435,
          histogram: (968, 1038, 1004, 1020, 1043, 1049, 1085, 1055, 1077, 1102, 1077, 1104, 1016, 1141, 1026, 969),
        ),
        FieldSnapshot(
          name: "debug[0]",
          min: -33,
          max: 45,
          unique: 75,
          histogram: (1936, 3360, 2413, 1737, 1282, 1004, 844, 693, 642, 568, 541, 477, 373, 356, 294, 254),
        ),
        FieldSnapshot(
          name: "debug[1]",
          min: -125,
          max: 214,
          unique: 141,
          histogram: (2061, 3251, 2580, 1998, 1631, 1159, 906, 702, 534, 392, 362, 301, 294, 224, 222, 157),
        ),
        FieldSnapshot(
          name: "debug[2]",
          min: -54,
          max: 33,
          unique: 83,
          histogram: (1684, 2803, 2474, 1983, 1642, 1230, 902, 707, 604, 544, 466, 435, 380, 329, 307, 284),
        ),
        FieldSnapshot(
          name: "debug[3]",
          min: 0,
          max: 0,
          unique: 1,
        ),
        FieldSnapshot(
          name: "motor[0]",
          min: 0,
          max: 2047,
          unique: 1172,
          histogram: (1017, 999, 1034, 981, 1050, 1026, 1033, 1022, 1030, 1069, 1075, 1066, 1133, 1045, 1112, 1082),
        ),
        FieldSnapshot(
          name: "motor[1]",
          min: 0,
          max: 2044,
          unique: 1193,
          histogram: (1078, 1056, 985, 1017, 983, 1077, 1030, 1067, 1009, 989, 1020, 1103, 1023, 1192, 1083, 1062),
        ),
        FieldSnapshot(
          name: "motor[2]",
          min: 0,
          max: 1901,
          unique: 1153,
          histogram: (1057, 1042, 1023, 1079, 1035, 1038, 979, 1029, 1048, 1058, 1051, 1037, 1096, 1126, 1069, 1007),
        ),
        FieldSnapshot(
          name: "motor[3]",
          min: 0,
          max: 2052,
          unique: 1190,
          histogram: (1032, 1067, 1018, 1055, 1030, 1063, 997, 1030, 1048, 1040, 1044, 1088, 1201, 1049, 1013, 999),
        ),
      ],
    ),
    slow: FrameSnapshot(
      fields: [
        FieldSnapshot(
          name: "flightModeFlags",
          min: 0,
          max: 524289,
          unique: 2,
          histogram: (1, 16773, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
        FieldSnapshot(
          name: "stateFlags",
          min: 0,
          max: 3,
          unique: 2,
          histogram: (1, 0, 0, 16773, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
        FieldSnapshot(
          name: "failsafePhase",
          min: 0,
          max: 0,
          unique: 1,
        ),
        FieldSnapshot(
          name: "rxSignalReceived",
          min: 0,
          max: 1,
          unique: 2,
          histogram: (1, 16773, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
        FieldSnapshot(
          name: "rxFlightChannelsValid",
          min: 0,
          max: 1,
          unique: 2,
          histogram: (1, 16773, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
      ],
    ),
  )),
]
